### PR TITLE
FileGlobs: Bypass the "files to ignore" question

### PR DESCRIPTION
**Pull request by marian-ald**
In CI mode there is no need to ask which files
to ignore, because the process is waiting for an
user to answer.

Closes: https://github.com/coala/coala-quickstart/issues/49

### CODE REVIEW

**Comment by hemangsk** [comment]
Hey @marian-ald, if we see FileGlobs.py, in line 33, there's an input being asked from the user
```
ignore_globs = ask_question(
            "Which files do you want coala to ignore?",
            printer=printer,
            typecast=list)
```
And in tests/Generation/FileGlobs.py we have,

```
31      with suppress_stdout(), simulate_console_inputs("ignore_dir/**"):
```
So the simulate_console_input()  sends input for this Question ("Which files do you want.."),
and since the input is "ignore_dir/**" , therefore all files in this directory should be ideally ignored.

But in the PR, you have non-interactive mode on by default. 
Therefore when tests execute... everything happens in non-interactive mode and so, the Question("files to ignore") is not asked and the files in ```ignore_dir``` directory aren't ignored as expected by the test.

File: coala_quickstart/coala_quickstart.py [hunk 1]
```diff
@@ -50,17 +50,20 @@ def main():
...
             default=project_dir,
             typecast=valid_path)
+        noninteractive = False
 
     project_files, ignore_globs = get_project_files(
         log_printer,
         printer,
-        project_dir)
+        project_dir,
+        noninteractive)
```
Comment (adtac) [inline coala_quickstart/coala_quickstart.py]: Just pass `args.non_interactive here` ;) no need for a `noninteractive` variable.

File: coala_quickstart/coala_quickstart.py [hunk 2]
```diff
@@ -50,17 +50,20 @@ def main():
     log_printer = LogPrinter(printer)
 
     project_dir = os.getcwd()
+    noninteractive = True
```
Comment (adtac) [inline coala_quickstart/coala_quickstart.py]: It shouldn't be non-interactive by default. It's interactive by default (that is, `noninteractive = False`). Anyway, see the  otehr comment - there's no need for this variable.

File: coala_quickstart/generation/FileGlobs.py [hunk 3]
```diff
@@ -16,6 +16,8 @@ def get_project_files(log_printer, printer, project_dir):
         A ``LogPrinter`` object.
     :param printer:
         A ``ConsolePrinter`` object.
+    :param noninteractive:
+        The running mode for coala-quickstart.
```
Comment (adtac) [inline coala_quickstart/generation/FileGlobs.py]: Maybe something like `Whether coala-quickstart is in non-interactive mode.` would be a better description? :)

File: coala_quickstart/generation/FileGlobs.py [hunk 4]
```diff
@@ -28,6 +30,12 @@ def get_project_files(log_printer, printer, project_dir):
                       color="green")
         ignore_globs = get_gitignore_glob(project_dir)
 
+    if noninteractive:
+        if ignore_globs is None:
+            file = open(".gitignore", "w")
+            file.write("src")
```
Comment (adtac) [inline coala_quickstart/generation/FileGlobs.py]: Why add `src` to the gitignore? This would ignore the source directory in some projects, which is the exact opposite of what we need!

Just create an empty file.

Comment (adtac) [inline coala_quickstart/generation/FileGlobs.py]: And you forgot to close the file.

Just do a `close(open(...))`.

File: coala_quickstart/generation/FileGlobs.py [hunk 5]
```diff
@@ -28,6 +30,12 @@ def get_project_files(log_printer, printer, project_dir):
                       color="green")
         ignore_globs = get_gitignore_glob(project_dir)
 
+    if noninteractive:
+        if ignore_globs is None:
+            file = open(".gitignore", "w")
```
Comment (adtac) [inline coala_quickstart/generation/FileGlobs.py]: You need to create this file in the *project directory*. Not the current directory.

This should be `os.path.join(project_dir, ".gitignore")`, not `".gitignore"`. See line 27 of this file for example.

File: tests/generation/FileGlobs.py [hunk 6]
```diff
@@ -30,11 +31,8 @@ def test_get_project_files(self):
         open(os.path.join("ignore_dir", "src.js"), "w").close()
 
         with suppress_stdout(), simulate_console_inputs("ignore_dir/**"):
-            res, _ = get_project_files(self.log_printer, self.printer, os.getcwd())
-            self.assertIn(os.path.join(os.getcwd(), "src", "file.c"), res)
-            self.assertIn(os.path.join(os.getcwd(), "root.c"), res)
-            self.assertNotIn(os.path.join(os.getcwd(), "ignore_dir/src.c"), res)
-            self.assertNotIn(os.path.join(os.getcwd(), "ignore_dir/src.js"), res)
+            res, _ = get_project_files(self.log_printer, self.printer, os.getcwd(),
+                False)
```
Comment (adtac) [inline tests/generation/FileGlobs.py]: This could be formatted better ;) Move all the arguments to the second line.

File: tests/generation/FileGlobs.py [hunk 7]
```diff
@@ -30,11 +31,8 @@ def test_get_project_files(self):
         open(os.path.join("ignore_dir", "src.js"), "w").close()
 
         with suppress_stdout(), simulate_console_inputs("ignore_dir/**"):
-            res, _ = get_project_files(self.log_printer, self.printer, os.getcwd())
-            self.assertIn(os.path.join(os.getcwd(), "src", "file.c"), res)
-            self.assertIn(os.path.join(os.getcwd(), "root.c"), res)
```
Comment (adtac) [inline tests/generation/FileGlobs.py]: Why did you remove these tests? These should still work when you give `noninteractive` as `False` to `get_project_files`, right?

File: tests/generation/FileGlobs.py [hunk 8]
```diff
@@ -15,6 +15,7 @@ class TestQuestion(unittest.TestCase):
     def setUp(self):
         self.printer = ConsolePrinter()
         self.log_printer = LogPrinter(self.printer)
+        self.noninteractive = True
```
Comment (adtac) [inline tests/generation/FileGlobs.py]: Nope, this shouldn't be a default. Add a test that tests the noninteractive mode by making sure there aren't any questions asked.

File: tests/generation/FileGlobs.py [hunk 9]
```diff
@@ -91,8 +89,8 @@ def test_get_project_files_gitignore(self):
         with suppress_stdout():
             self.assertEqual(
                 sorted(get_project_files(
-                    self.log_printer, self.printer, os.getcwd())[0]),
-                sorted(files))
+                    self.log_printer, self.printer, os.getcwd(),
+                self.noninteractive)[0]), sorted(files))
```
Comment (adtac) [inline tests/generation/FileGlobs.py]: Add a separate test for non-interactive mode. Don't touch the old tests because they shouldn't really be affected.

File: coala_quickstart/generation/FileGlobs.py [hunk 10]
```diff
@@ -7,7 +7,7 @@
 from coalib.collecting.Collectors import collect_files
 
 
-def get_project_files(log_printer, printer, project_dir):
+def get_project_files(log_printer, printer, project_dir, noninteractive):
```
Comment (adtac) [inline coala_quickstart/generation/FileGlobs.py]: Make `False` as the default value for `noninteractive`. That way the older tests will work as before. Then add a new test just to test `noninteractive=True`.

And maybe `non_interactive` is better than `noninteractive`? Since we use `args.non_interactive` ;)

File: coala_quickstart/generation/FileGlobs.py [hunk 11]
```diff
@@ -28,6 +30,12 @@ def get_project_files(log_printer, printer, project_dir):
                       color="green")
         ignore_globs = get_gitignore_glob(project_dir)
 
+    if noninteractive:
+        if ignore_globs is None:
+            file = open(".gitignore", "w")
+            file.write("src")
+            ignore_globs = get_gitignore_glob(project_dir)
+
     if ignore_globs is None:
```
Comment (adtac) [inline coala_quickstart/generation/FileGlobs.py]: Actually, the best way to solve this is to change this if as:

```
if ignore_globs is None:
    ...
elif not non_interactive:
    ignore_globs = []  # since ignore_globs is basically a list in the end, let's create an empty list
```

That way we don't even need to create an empty gitignore.

Comment (marian-ald) [inline coala_quickstart/generation/FileGlobs.py]: @adtac I think also it's ok like this
`if ignore_globs is None:
    if non_interactive:
        ignore_globs = [ ]` 
Like you said, we don't need to create a file. And when we check
`
    if ignore_globs is None:
       ...
            "Which files do you want coala to ignore?",
`
It won't ask the question :smile:

Comment (adtac) [inline coala_quickstart/generation/FileGlobs.py]: yeah, that looks good too :+1:

File: tests/generation/FileGlobs.py (hunk 9 above)
Comment (marian-ald) [inline tests/generation/FileGlobs.py]: @adtac You mean don't touch at all this test? But when I do a pytest I get " missing 1 required positional argument", because my function has 4 arguments.

File: coala_quickstart/generation/FileGlobs.py (hunk 10 above)
Comment (marian-ald) [inline coala_quickstart/generation/FileGlobs.py]: This is a little unclear to me... I mean these tests :(. Why to make non_interactive False here, in FileGlobs? I receive it in the definition of the function. Correct me if I'm wrong...

File: tests/generation/FileGlobs.py [hunk 12]
```diff
@@ -91,8 +92,8 @@ def test_get_project_files_gitignore(self):
...
             self.assertEqual(
                 sorted(get_project_files(
-                    self.log_printer, self.printer, os.getcwd())[0]),
-                sorted(files))
+                    self.log_printer, self.printer, os.getcwd(),
+                self.non_interactive)[0]), sorted(files))
 
         os.remove(".gitignore")
-        os.chdir(orig_cwd)
+        os.chdir(orig_cwd)
```
Comment (jayvdb) [inline tests/generation/FileGlobs.py]: EOL at EOF please

File: tests/generation/FileGlobs.py [hunk 13]
```diff
@@ -30,7 +30,8 @@ def test_get_project_files(self):
         open(os.path.join("ignore_dir", "src.js"), "w").close()
 
         with suppress_stdout(), simulate_console_inputs("ignore_dir/**"):
-            res, _ = get_project_files(self.log_printer, self.printer, os.getcwd())
+            res, _ = get_project_files(
```
Comment (jayvdb) [inline tests/generation/FileGlobs.py]: why ignore the tuple's second item?
Can't you check that value ?

**Comment by hemangsk** [comment]
Hey @marian-ald 😊 , Are you planning to work on this? 
If not, do you mind if I take this up from here? Its very crucial for  #90

**Comment by jayvdb** [comment]
Completed in https://github.com/coala/coala-quickstart/pull/94
